aten/src/ATen/core/type.cpp

```diff
@@ -523,6 +523,25 @@ TensorTypePtr TensorType::merge(TensorTypePtr other, bool merge_sizes) const {
       undef);
 }
 
+template <typename T>
+bool is_null_or_equal(c10::optional<T> a, c10::IntArrayRef b) {
+  return !a.has_value() || a.value() == b;
+}
+
+bool TensorType::matchTensor(const at::Tensor& t) {
+  bool undef = undefined().value_or(true);
```

Comment:
why isn't it `value_or(!t.defined())`?

Reply:
They are logically equivalent ;) : 
`if (undefined().value_or(!t.defined()) == !t.defined())`

Reply:
And we want to keep `undef` as `status` of `TensorType` without mixing `t`'s state. So that it can be reused in the return statement as well.